week03/1-to-50/src/utils/gameLogic.js

```diff
@@ -0,0 +1,88 @@
+import { useState, useEffect } from 'react';
+import { LEVELS } from '../constants/level';
+
+export const useGameLogic = (gameLevel) => {
+  const { firstSet, secondSet } = LEVELS[gameLevel];
+  const [firstCards, setFirstCards] = useState([]);
+  const [secondCards, setSecondCards] = useState([]);
+  const [displayCards, setDisplayCards] = useState([]);
+  const [nextNumber, setNextNumber] = useState(1);
+  const [timer, setTimer] = useState(0);
+  const [isTimerStarted, setIsTimerStarted] = useState(false);
+
+  const generateShuffledCards = () => {
+    const firstNumbers = Array.from(
+      { length: firstSet },
+      (_, index) => index + 1
+    );
+
+    const secondNumbers = Array.from(
+      { length: secondSet - firstSet },
+      (_, index) => index + firstSet + 1
+    );
+
+    const shuffledFirstNumbers = firstNumbers.sort(() => Math.random() - 0.5);
+    const shuffledSecondNumbers = secondNumbers.sort(() => Math.random() - 0.5);
+
+    return { shuffledFirstNumbers, shuffledSecondNumbers };
+  };
+
+  const initializeGame = () => {
+    const { shuffledFirstNumbers, shuffledSecondNumbers } =
+      generateShuffledCards();
+    setFirstCards(shuffledFirstNumbers);
+    setSecondCards(shuffledSecondNumbers);
+    setDisplayCards(shuffledFirstNumbers);
+    setNextNumber(1);
+    setTimer(0);
+    setIsTimerStarted(false);
+  };
+
+  const handleCardClick = (number, index, initiateGame) => {
+    if (!isTimerStarted) {
+      initiateGame(); // 타이머 시작을 외부에서 호출
+      setIsTimerStarted(true);
+    }
+
+    if (number !== nextNumber) {
+      alert(`${nextNumber}을 클릭해주세요.`);
+      return;
+    }
+
+    if (number === secondSet) {
+      alert(`게임 종료! 걸린 시간: ${timer.toFixed(2)}초`);
+      initializeGame();
+      return;
+    }
+
+    const newDisplayCards = [...displayCards];
+    if (number <= firstSet) {
+      newDisplayCards[index] = secondCards[index];
+    } else {
+      newDisplayCards[index] = ''; // 빈 값으로 설정하여 카드 숨기기
+    }
+    setDisplayCards(newDisplayCards);
+    setNextNumber((prev) => prev + 1);
+  };
+
+  useEffect(() => {
+    const interval = setInterval(() => {
+      if (isTimerStarted) {
+        setTimer((prev) => prev + 0.01);
+      }
+    }, 10);
+    return () => clearInterval(interval);
+  }, [isTimerStarted]);
+
+  useEffect(() => {
+    initializeGame();
+  }, [gameLevel]);
```

Comment:
p3: react 공식문서에 의하면

> When you’re not sure whether some code should be in an Effect or in an event handler, ask yourself *why* this code needs to run. Use Effects only for code that should run *because* the component was displayed to the user.

-> **사용자에게 해당 컴포넌트가 보여졌다는 이유만으로도 호출되어야 하는 코드는 useEffect를 사용하고, 사용자와 인터랙션이 있을 때는 useEffect를 사용하지 않는 편이 좋다.**
라고 합니다! 
저도 useEffect 를 자주종종 쓰긴 하지만 참고용으로 언급 드려요!